### PR TITLE
PowerDNS recursor check

### PR DESCRIPTION
**Pull request by AntoCard**
Check for PowerDNS recursor metrics. Needs pdns-recursor experimental web server active.

### CODE REVIEW

File: checks.d/powerdns-recursor.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Datadog
+from checks import AgentCheck
+
+# Other
+import json
```
Comment (remh) [inline checks.d/powerdns-recursor.py]: You should import simplejson instead, it's more efficient

File: checks.d/powerdns-recursor.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import urllib2
```
Comment (remh) [inline checks.d/powerdns-recursor.py]: You should use the requests library instead.

File: checks.d/powerdns-recursor.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+class PdnsChecks(AgentCheck):
+
+
+    def _get_pdns_stats(self, instance):
+        config = self._get_config(instance)
+        request = urllib2.Request(config.pdns_url + ':' + str(config.port) + "/servers/localhost/statistics", headers={"X-API-Key" : config.api_key})
+        content = urllib2.urlopen(request).read()
+        result = json.loads(content)
+        return result
```
Comment (remh) [inline checks.d/powerdns-recursor.py]: Do you have an example of a result ? How many stats would be returned most of the time ?

**Comment by remh** [comment]
Thanks a lot @AntoCard for this!
According to this page https://doc.powerdns.com/md/recursor/stats/ there are a lot of metrics that can be collected.

I think we should explicitly list the metrics that we want to collect (it will be also easier to send them with the more appropriate type).

Thanks again!

**Comment by irabinovitch** [comment]
@AntoCard I wanted to echo Remi's thanks here too.  This looks great.  Lets sync up on what metrics make sense to pull in by default, and maybe collaborate on a readme for setup so that folks know what the requirements/dependencies are.

**Comment by irabinovitch** [comment]
@AntoCard I wanted to echo Remi's thanks here too.  This looks great.  Lets sync up on what metrics make sense to pull in by default, and maybe collaborate on a readme for setup so that folks know what the requirements/dependencies are.

**Comment by AntoCard** [comment]
Hi guys, thanks for the feedback. It's really useful. I'll work in improving this and update my PR. Cheers.

**Comment by irabinovitch** [comment]
@AntoCard  Have you had a chance to review the feedback? We'd love to include this in an upcoming release.

**Comment by AntoCard** [comment]
Hi @irabinovitch, I'm going to collect a list of the metrics we find useful here in Notonthehighstreet, maybe we can use some of them as default metrics.

**Comment by irabinovitch** [comment]
@AntoCard Sounds good. Looking forward to it!

**Comment by irabinovitch** [comment]
@AntoCard Were you able to look at collecting a list of metrics? We're hoping for  a short window between 5.7 and 5.8, so would be great to get this  addressed and included early.

**Comment by AntoCard** [comment]
@irabinovitch this are the queries we use more often:

`pdns.answers0_1` Queries answered within 1 millisecond
`pdns.answers1_10` Queries answered within 10 milliseconds
`pdns.answers10_100` Queries answered within 100 milliseconds
`pdns.answers100_1000` Queries answered within 1 second
`pdns.answers_slow` Queries answered after 1 second
`pdns.outgoing_timeouts` Number of timeouts on outgoing UDP queries
`pdns.noerror_answers` Number of times it answered NOERROR
`pdns.servfail_answers` Number of times it answered SERVFAIL

**Comment by janeczku** [comment]
This PR is useful!
Here is a basic set of the most useful metrics reported by PowerDNS recursor:

**Gauges**

`concurrent-queries`
`cache-entries`
`qa-latency` (usec)

**Counters**

`questions`
`tcp-questions`
`all-outqueries`
`tcp-outqueries`
`servfail-answers`
`nxdomain-answers`
`noerror-answers`

**Histograms**

Would be nice if the PR could be extended to derive histogram metrics from these cumulative histogram counters. Otherwise we can of course just submit them as counters.

`answers0-1` (0-1ms latency)
`answers1-10` (1-10ms latency)
`answers10-100` (10-100ms latency)
`answers100-1000` (100-1000ms latency)
`answers-slow` (>1000ms latency)

**Comment by janeczku** [comment]
@AntoCard :laughing: you beat me to it by a minute!

**Comment by AntoCard** [comment]
@janeczku your list is great! Are you using it in a Docker/Mesos environment?

**Comment by janeczku** [comment]
Almost. I am running multiple instances of dnsdist and PowerDNS in a Docker/Rancher environment.

File: conf.d/powerdns-recursor.yaml.j2 [hunk 4]
```diff
@@ -0,0 +1,6 @@
+init_config:
```
Comment (remh) [inline conf.d/powerdns-recursor.yaml.j2]: Nitpick but you should remove the `.j2` file extension.

**Comment by remh** [comment]
Thanks a lot guys! Do you think you could add a test similar to our other integrations tests ? It can be mocked if it's not easy to set up powerdns

**Comment by remh** [comment]
Also, we should likely only include the metrics that are relevant. It's better than collecting metrics that are not useful

**Comment by degemer** [comment]
Closing in favor of https://github.com/DataDog/dd-agent/pull/2490

**Pull request closed without merging**